File: weekly/344_/q2.cc
class FrequencyTracker {
public:
    FrequencyTracker() {}
    
    void add(int number) {
        auto c = ++cnts[number];
        --freq[c - 1];
        ++freq[c];
    }
    
    void deleteOne(int number) {
        auto c = cnts[number];
        if (c == 0) return;
        --freq[c];
        if (c != 1) ++freq[c - 1];
        --cnts[number];
    }
    
    bool hasFrequency(int frequency) {
        return freq[frequency];
    }

private:
    map<int, int> cnts;
    map<int, int> freq;
};

/**
 * Your FrequencyTracker object will be instantiated and called as such:
 * FrequencyTracker* obj = new FrequencyTracker();
 * obj->add(number);
 * obj->deleteOne(number);
 * bool param_3 = obj->hasFrequency(frequency);
 */

File: weekly/344_/go-solutions/q2.go
type FrequencyTracker struct {
    freq map[int]int
    cnt map[int]int
}


func Constructor() FrequencyTracker {
    return FrequencyTracker {
        freq: map[int]int{},
        cnt: map[int]int{},
    }
}


func (this *FrequencyTracker) Add(number int)  {
    if this.cnt[number] > 0 {
        this.freq[this.cnt[number]]--
    }
    this.cnt[number]++
    this.freq[this.cnt[number]]++
}


func (this *FrequencyTracker) DeleteOne(number int)  {
    if this.cnt[number] == 0 {
        return
    }
    this.freq[this.cnt[number]]--
    this.cnt[number]--
    if this.cnt[number] > 0 {
        this.freq[this.cnt[number]]++
    }
}


func (this *FrequencyTracker) HasFrequency(frequency int) bool {
    return this.freq[frequency] > 0
}


/**
 * Your FrequencyTracker object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Add(number);
 * obj.DeleteOne(number);
 * param_3 := obj.HasFrequency(frequency);
 */

